### PR TITLE
TM region

### PR DESCRIPTION
**Pull request by jaredweiss**
Basic initial implementation of a TM region. 

Note: Needed to make a few small changes (recasting variables, etc.) to the actual temporal memories  to get this to play nicely with CPP. These changes should be carefully reviewed at some point.

### CODE REVIEW

**Comment by jaredweiss** [comment]
@scottpurdy @chetan51 
not sure if you've checked this out yet. i think it's a good working implementation, but there's still some work to be done. would appreciate feedback nonetheless

**Comment by scottpurdy** [comment]
@jaredweiss - I hadn't seen this.

@chetan51 - Can you review the region implementation? Do you think we should add this to NuPIC or keep in research for now?

File: sensorimotor/sensorimotor/TMRegion.py [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+from nupic.regions.PyRegion import PyRegion
+
+from sensorimotor.general_temporal_memory import GeneralTemporalMemory
+from sensorimotor.fast_general_temporal_memory import FastGeneralTemporalMemory
+from nupic.research.monitor_mixin.temporal_memory_monitor_mixin import (
+  TemporalMemoryMonitorMixin)
+
+
+
+class MonitoredFastGeneralTemporalMemory(TemporalMemoryMonitorMixin,
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Can we include regular `TemporalMemory`?

File: sensorimotor/sensorimotor/TMRegion.py [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+    return GeneralTemporalMemory
+  elif tmImp == "fast":
+    return FastGeneralTemporalMemory
+  elif tmImp == "generalMonitored":
+    return MonitoredGeneralTemporalMemory
+  elif tmImp == "fastMonitored":
+    return MonitoredFastGeneralTemporalMemory
+  else:
+    raise RuntimeError("Invalid temporal memory implementation '{imp}'. "
+                       "Legal values are: 'general' and 'fast'"
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: This list is incomplete

File: sensorimotor/sensorimotor/TMRegion.py [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+
+class MonitoredFastGeneralTemporalMemory(TemporalMemoryMonitorMixin,
+                                         FastGeneralTemporalMemory): pass
+class MonitoredGeneralTemporalMemory(TemporalMemoryMonitorMixin,
+                                      GeneralTemporalMemory): pass
+
+
+
+def getDefaultTMImp():
+  """
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Move docstring to 1 line

File: sensorimotor/sensorimotor/TMRegion.py [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+  Return the default temporal memory implementation for this region.
+  """
+  return "fast"
+
+
+
+def getTMClass(tmImp):
+  """ Return the class corresponding to the given spatialImp string
+  """
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Move docstring to 1 line

File: sensorimotor/sensorimotor/TMRegion.py [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+      constraints="bool"),
+  )
+
+  return tmSpec, otherSpec
+
+
+
+class TMRegion(PyRegion):
+
+  """
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Add docstring

File: sensorimotor/sensorimotor/TMRegion.py [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+               columnDimensions=(2048,),
+               cellsPerColumn=32,
+               activationThreshold=13,
+               initialPermanence=0.21,
+               connectedPermanence=0.50,
+               minThreshold=10,
+               maxNewSynapseCount=20,
+               permanenceIncrement=0.10,
+               permanenceDecrement=0.10,
+               predictedSegmentDecrement = 0.0,
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Remove extra spaces around `=`

File: sensorimotor/sensorimotor/TMRegion.py [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+               seed=42,
+               learnOnOneCell=False,
+               tmImp=getDefaultTMImp(),
+               **kwargs):
+    # Pull out the pooler arguments automatically
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Rename "pooler"

File: sensorimotor/sensorimotor/TMRegion.py [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+    # These calls whittle down kwargs and create instance variables of TMRegion
+    self._tmClass = getTMClass(tmImp)
+    tmArgTuples = _buildArgs(self._tmClass, self, kwargs)
+
+    # Make a list of automatic pooler arg names for later use
+    self._tmArgNames = [t[0] for t in tmArgTuples]
+
+    # Defaults for all other parameters
+    self.columnDimensions = copy.deepcopy(columnDimensions)
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Why is copy needed here?

File: sensorimotor/sensorimotor/TMRegion.py [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+                                   title="pa-{name}".format(name=name))
+
+
+  @classmethod
+  def getBaseSpec(cls):
+    """Return the base Spec for TMRegion.
+
+    Doesn't include the tm parameters
+    """
+    spec = dict(
```
Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: This doesn't seem very DRY considering https://github.com/numenta/nupic.research/pull/247/files#diff-482d0f1f30e71b5dc3b34480ed6d72fdR167

File: sensorimotor/sensorimotor/fast_general_temporal_memory.py [hunk 10]
```diff
@@ -88,8 +88,7 @@ def burstColumns(self,
     activeCells = set()
     winnerCells = set()
     learningSegments = set()
-
-    unpredictedColumns = activeColumns - predictedColumns
+    unpredictedColumns = set(activeColumns) - predictedColumns
```
Comment (chetan51) [inline sensorimotor/sensorimotor/fast_general_temporal_memory.py]: Why was this necessary?

File: sensorimotor/sensorimotor/general_temporal_memory.py [hunk 11]
```diff
@@ -401,7 +401,7 @@ def learnOnApicalSegments(self,
     """
     for winnerCell in winnerCells:
       winnerSegments = connections.segmentsForCell(winnerCell)
-      if len(winnerSegments & (prevActiveSegments | learningSegments)) == 0:
+      if len(set(winnerSegments) & (prevActiveSegments | learningSegments)) == 0:
```
Comment (chetan51) [inline sensorimotor/sensorimotor/general_temporal_memory.py]: Why was this necessary?

**Comment by chetan51** [comment]
Review complete.

File: sensorimotor/sensorimotor/TMRegion.py (hunk 1 above)
Comment (jaredweiss) [inline sensorimotor/sensorimotor/TMRegion.py]: Would it make sense to write a regular `TMRegion` in nupic and then just change this one to be a `GeneralTMRegion` to extend the regular one and have it in nupic.research? Or should this one just work for both?

Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: That might be overly complicated. To be consistent with `TPRegion`, and since all the `TemporalMemory` subclasses have the same interface, we can just have this one work for all of them.

Comment (jaredweiss) [inline sensorimotor/sensorimotor/TMRegion.py]: Got it!

File: sensorimotor/sensorimotor/TMRegion.py (hunk 8 above)
Comment (jaredweiss) [inline sensorimotor/sensorimotor/TMRegion.py]: columnDimensions will be passed as type `numpy.ndarray` so it'll disappear if it's not explicitly copied like this. Another option would be to recast it as a python array instead of copying the ndarray.

File: sensorimotor/sensorimotor/TMRegion.py (hunk 9 above)
Comment (jaredweiss) [inline sensorimotor/sensorimotor/TMRegion.py]: I'm going to change this when I include the nupic TM and fastTM

File: sensorimotor/sensorimotor/fast_general_temporal_memory.py (hunk 10 above)
Comment (jaredweiss) [inline sensorimotor/sensorimotor/fast_general_temporal_memory.py]: I think when I was first writing this I was having some serialization issues. Shouldn't be a problem anymore though so I'll remove it.

**Comment by scottpurdy** [comment]
@chetan51 - Has @jaredweiss responded to all of your feedback?

**Comment by chetan51** [comment]
It sounds like he's still working on including regular `TemporalMemory`.

**Comment by jaredweiss** [comment]
@chetan51 @scottpurdy Yep! Tiny bit more complicated than I originally thought since the inputs/outputs for the different types of TM are slightly different but should have it done soon anyway. Its not that tough.

**Comment by jaredweiss** [comment]
Also was trying to work on serialization so that the network could be saved but got a bit hung up on how to do that. Not really sure how to go forward with that atm

**Comment by chetan51** [comment]
Also once this is finished, I think we should put it in nupic.

File: sensorimotor/sensorimotor/TMRegion.py (hunk 9 above)
Comment (jaredweiss) [inline sensorimotor/sensorimotor/TMRegion.py]: What did you mean by this? Were you referring to the fact that all of the arguments are specified in the spec rather than being parsed from the `__init__` method? Since there's no actual arguments actually specified in the `__init__` for the general temporal memories (other than `*args` and `**kwargs`) I'm not really sure what'd be the best way to populate the spec. 

`getArgumentDescriptions(tmClass.__init__)` looks through the `__init__` args so I'd be in favor of just replacing the `*args`/`**kwargs** with the actual parameter names and let that function do the work of populating the spec. Would that be a better option than manually defining the majority of the spec?

Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: Ah, looking at my comment again, I realized that these parameters are not the same as the ones specified in https://github.com/numenta/nupic.research/pull/247/files#diff-482d0f1f30e71b5dc3b34480ed6d72fdR167. You can ignore my original comment, but if you think you have a cleaner way to do this, feel free!

Comment (jaredweiss) [inline sensorimotor/sensorimotor/TMRegion.py]: Aah. Ok. 

I think there's two possible ways to do this. One is to get rid of the `*args`/`**kwargs` in the `__init__` constructors for `GeneralTemporalMemory`, `FastGeneralTemporalMemory`, etc. and replace them with explicitly defined parameters but then any changes in the underlying base implantation of the default parameters would need to be copied into each of the classes. 

I think it'd make more sense to just write the `getSpec` (and the helper function `getAdditionalSpec`) to just return the proper parameters for each of the implementations of TM.

Comment (chetan51) [inline sensorimotor/sensorimotor/TMRegion.py]: It's not necessary to make those changes to get this PR merged. Let me know what you want to do.

**Comment by subutai** [comment]
Is this PR still valid? It needs to be updated to the new structure before we can merge

**Comment by scottpurdy** [comment]
@chetan51 - Is this still valid? We should wait for serialization (just raise an exception if the serialization fucntions are called). Or does it make more sense to close this in favor of creating the region in nupic?

**Comment by chetan51** [comment]
This is a good start on a TM region, so maybe we should just merge it to nupic.research now, and add serialization and make it ready for nupic later?

**Comment by scottpurdy** [comment]
Replacing with #311 that builds off this branch and updates to new directory structure.
